### PR TITLE
Improve tiller resource so it handles broken tenant clusters

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#6561

### CODE REVIEW

File: pkg/v18/resource/tiller/resource.go [hunk 1]
```diff
@@ -125,26 +125,26 @@ func (r *Resource) ensureTillerInstalled(ctx context.Context, obj interface{}) e
 		r.logger.LogCtx(ctx, "level", "debug", "message", "timeout fetching certificates")
 
 		// A timeout error here means that the cluster-operator certificate
-		// for the current guest cluster was not found. We can't continue
+		// for the current tenant cluster was not found. We can't continue
```
Comment (rossf7) [inline pkg/v18/resource/tiller/resource.go]: Some tidying up to fix the naming.

File: pkg/v18/resource/tiller/resource.go [hunk 2]
```diff
@@ -125,26 +125,26 @@ func (r *Resource) ensureTillerInstalled(ctx context.Context, obj interface{}) e
...
 		// without a Helm client. We will retry during the next execution, when
 		// the certificate might be available.
 		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation")
 		reconciliationcanceledcontext.SetCanceled(ctx)
 
 		return nil
-	} else if helmclient.IsTillerInstallationFailed(err) {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "Tiller installation failed")
+	} else if helmclient.IsTillerNotFound(err) {
```
Comment (rossf7) [inline pkg/v18/resource/tiller/resource.go]: The error we matched on before wasn't even used. 🤦‍♂

File: service/service.go [hunk 3]
```diff
@@ -164,6 +164,10 @@ func New(config Config) (*Service, error) {
 			Logger:        config.Logger,
 
 			CertID: certs.ClusterOperatorAPICert,
+			// This is used by the Tiller resource where we use a shorter max
+			// wait because the tenant cluster is unavailable. The retry
+			// resource is enabled so we retry once.
+			EnsureTillerInstalledMaxWait: 15 * time.Second,
```
Comment (rossf7) [inline service/service.go]: Previously we used the default 60 seconds and retried which was WAY too long.

File: pkg/v18/resource/tiller/resource.go (hunk 2 above)
Comment (tomahawk28) [reply pkg/v18/resource/tiller/resource.go]: How did you find that? can we have a unittest for this?

**Comment by rossf7** [comment]
OK fixing the build error. I'm also going to backport the fix to match the correct error to all versions.

Otherwise I worry the operatorkit errors alert will fire. On it.

File: pkg/v18/resource/tiller/resource.go (hunk 2 above)
Comment (rossf7) [reply pkg/v18/resource/tiller/resource.go]: I found it during manual testing. I'm not sure how we can catch that in a unit test.

File: pkg/v10/resource/chartoperator/create.go [hunk 4]
```diff
@@ -37,7 +37,7 @@ func (r *Resource) ApplyCreateChange(ctx context.Context, obj, createChange inte
 				resourcecanceledcontext.SetCanceled(ctx)
 
 				return nil
-			} else if helmclient.IsTillerInstallationFailed(err) {
+			} else if helmclient.IsTillerNotFound(err) {
```
Comment (rossf7) [inline pkg/v10/resource/chartoperator/create.go]: This is all renaming to use the error that is actually used. Except for reducing the backoff which I'll also highlight.

File: service/service.go [hunk 5]
```diff
@@ -164,6 +164,10 @@ func New(config Config) (*Service, error) {
 			Logger:        config.Logger,
 
 			CertID: certs.ClusterOperatorAPICert,
+			// This is used by the Tiller resource where we use a shorter max
+			// wait because the tenant cluster may be unavailable. If so the
+			// reconciliation loop is cancelled and we retry in the next loop.
+			EnsureTillerInstalledMaxWait: 30 * time.Second,
```
Comment (rossf7) [inline service/service.go]: This is where the new max wait is configured.

File: service/service.go [hunk 6]
```diff
@@ -164,6 +164,12 @@ func New(config Config) (*Service, error) {
 			Logger:        config.Logger,
 
 			CertID: certs.ClusterOperatorAPICert,
+			// TODO: Reduce the max wait to reduce delay when processing
+			// broken tenant clusters.
+			//
+			// See https://github.com/giantswarm/giantswarm/issues/6703.
```
Comment (xh3b4sd) [inline service/service.go]: I think the TODO issue should be put in some board. Do you have one? 

```suggestion
			//     https://github.com/giantswarm/giantswarm/issues/6703
```

File: Gopkg.toml [hunk 7]
```diff
@@ -104,7 +104,7 @@ required = [
   name = "github.com/giantswarm/operatorkit"
 
 [[constraint]]
-  branch = "master"
+  branch = "improve-error-handling"
```
Comment (xh3b4sd) [inline Gopkg.toml]: The vendor deps needs updating still.

File: service/service.go (hunk 6 above)
Comment (rossf7) [reply service/service.go]: Yes I've added to the batman board and my own.

This is nicer but I copied `fmt` 😂  I'll update it!

https://github.com/giantswarm/fmt/blob/master/go/todo_comments.md

**Comment by rossf7** [comment]
Tested on `ginger`.

**Pull request merged**